src/PROPOSAL/detail/PROPOSAL/Propagator.cxx

```diff
@@ -139,93 +142,112 @@ Interaction::Loss Propagator::DoStochasticInteraction(ParticleState& p_cond,
+                for (auto& r: random_numbers) {
+                    r = rnd_generator();
+                }
+                Logging::Get("proposal.propagator")->debug("Discard current set of random numbers for multiple"
+                                                           " scattering since we are stuck in a loop.");
+                grammage = grammage_next_interaction;
+                energy = energy_next_interaction;
+                distance = density->Correct(state.position, state.direction, grammage, max_distance);
             }
+        } else {
+            throw std::logic_error("Error in AdvanceParticle: We shouldn't end up here!");
+        }
 
-            AdvanceDistance[ReachedBorder] = CalculateDistanceToBorder(
-                state.position, mean_direction, *geometry);
-            advancement_type = minimize(AdvanceDistance);
-            control_distance = AdvanceDistance[advancement_type];
-        } while (std::abs(advance_distance - control_distance)
-            > PARTICLE_POSITION_RESOLUTION);
-        advance_distance = control_distance;
-        advance_grammage = density->Calculate(
-            state.position, mean_direction, advance_distance);
-        new_position = state.position + advance_distance * mean_direction;
-    }
+        // Calculate scattering proposal
+        std::tie(mean_direction, new_direction) = utility.DirectionsScatter(
+                grammage, state.energy, energy, state.direction, rnd);
+
+        // Check step
+        double distance_to_border = CalculateDistanceToBorder(state.position, mean_direction, *geometry);
+        bool is_inside = geometry->IsInside(state.position, mean_direction);
+        if (!is_inside) {
+            // Special case: We are on the sector border, but scattering back outside the current sector!
+            // Update sector and recalculate values
+            advancement_type = InvalidStep;
+            auto new_sector = GetCurrentSector(state.position, mean_direction);
+            utility = get<UTILITY>(new_sector);
+            density = get<DENSITY_DISTR>(new_sector);
+            geometry = get<GEOMETRY>(new_sector);
+            grammage_next_interaction = utility.LengthContinuous(state.energy, energy_next_interaction);
+            energy = energy_next_interaction;
+            distance = -1;
+            grammage = -1;
+        } else if (distance <= distance_to_border && distance <= max_distance && energy == energy_next_interaction) {
+            // reached interaction
+            advancement_type = ReachedInteraction;
+        } else if (distance <= distance_to_border && distance == max_distance) {
+            // reached max distance
+            advancement_type = ReachedMaxDistance;
+        } else if (std::abs(distance - distance_to_border) <= PARTICLE_POSITION_RESOLUTION) {
+            // reached geometry border
+            advancement_type = ReachedBorder;
+            distance = distance_to_border;
+        } else {
```

Comment:
When do we run into this case, only if the distance is set to INF due to a DensityError?
Or what other cases exist?

Reply:
You mean the last else bracket?

As indicated by the comment, this means that we didn't finish our iteration: With the current combination of our propagation step length and the scattering angle, we do not hit the sector border (we either overshoot the border, or we undershoot it). Therefore we have to perform another iteration step until we have reached the sector border within our particle precision (... or until we reached the maximum propagation distance, or reached an interaction, which are the other cases that can appear)